### ISSUE TITLE
[Permissions] Homepage manager role needs the ability to create FW Banners & edit Path field

### ISSUE THREAD

**Issue by FranECross** [department-of-veterans-affairs/va.gov-cms#18928]
## Status
**[2024-08-15] [Fran]** Dave P completed his review and recommendation is to add the ability to create a Full Width Banner to the existing Home Page Manager role (which Josh already has). Slack thread [here](https://dsva.slack.com/archives/C52CL1PKQ/p1723680168114699) providing info to Michelle M.
**[2024-08-14] [Fran]** Dave P will look into roles; another ticket will be created if drupal new role is needed. Does not need to be implemented in the upcoming sprint if a new role is needed.

## Description
Josh Tuscher has a Content Publisher role, and in that role, he can modify and archive full width banners, but can't create them.

Randi Hecht has confirmed that she would expect OPIA to be able to create new Full Width Banners. [Slack convo](https://dsva.slack.com/archives/C07GMJHAUM8/p1723508874024699)

Consider if there is an existing role that can be assigned, without providing Administrator permissions, or if a new role needs to be created.

## Additional Information
- Jill confirmed this via Tugboat
- Jill confirmed that a content publisher, in the VACO section, which includes OPIA Alerts and Promo banners section, SHOULD have permissions to create new Full Width Banners (Slack convo.)
- Jill confirmed that at this time, Administrators can create new FWBs

**Regarding Permissions of Full Width Banners and the interplay of Sections:**
Jill  got the Permissions part of Full width banner permissions to load, and can see that Content Publishers are not generally allowed to create new FWBs. BUT: I don't really understand the interplay of Sections, and how they affect role permissions on content type actions. So a Drupal eng will need to speak to that.
(Took a shot of Vet Center permissions as a contrast that sort of helps see that Sections is doing a lot for us in this realm, I think.)

<img width="1449" src="https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBOUpVQmc9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--6d915f59292cae1c1e62445ae411e2869cf8bfd4/Screenshot%202024-08-13%20112548.png" alt="Screenshot 2024-08-13 112548.png" />

#### Summary (from Jill) of the info below: 

A content type has permissions for each role. A user has roles. A user has Workbench assignments that correlate to a section. And if a user has the correct role, and Workbench section for the content they're interested in: they can do stuff defined by the role permissions. 

#### Info from Ian Sears:
- Sections (Taxonomy) in the UI lives [here](https://prod.cms.va.gov/admin/structure/taxonomy/manage/administration/overview)
- When we Edit the Term... (the Taxonomy section) we are able to choose the tab "Section members"
One of the Section Member is the Workbench Group created as a candidate for this Term-Edit_SectionMember area.
- As you know and understand, Roles restrict permissions. In our CMS we also use a concept/module called WorkBench. What we've done is break out all the VAMC Systems, and each Vet Center (just examples here) as individual things in WorkBench. You can think of these kind of like WorkBench "Sections". Each of these... even further-restricts privileges and permissions.
At the end of the day, we provide you a ROLE enabling you to Edit VAMC..."stuff" and then enable that in a specific WorkBench Section, like VA Tampa health care, or  Watertown Vet Center. To see this in action, take a look at [this](https://prod.cms.va.gov/user/5117/workbench_access) which is the Workbench access for the user Zachary Hughes

### User story

**AS A** Content Publisher in the VACO section
**I WANT** the ability to create Full Width Banners
**SO THAT** I don't rely on CAIA to create them for me, and so that I'm not tempted to simply edit an archived one, which is problematic in that any user who dismissed the banner previously won't see this 'edited' version of the archived / republished banner.

### Engineering notes / background
Pull request is up that adds the permission to create FWB to the Homepage manager role: https://github.com/department-of-veterans-affairs/va.gov-cms/pull/18958 

However: the Path field is not editable for a Homepage manager. Looks like that is controlled by a custom module https://github.com/department-of-veterans-affairs/va.gov-cms/blob/fe080b2efb86459900e581ea0974895c606f003c/docroot/modules/custom/va_gov_banner/src/EventSubscriber/EntityEventSubscriber.php

The PR will need an update to allow Homepage Managers to modify the Path field when creating new Full Width Banners (/node/add/banner)

### Quality / testing notes
In lower environments/Tugboat, you can login as Joshua.Tuscher@va.gov using the global lower envs password, to test.

## Acceptance criteria

- [ ] Users assigned the role Homepage Manager are able to create Full Width Banners and populate required fields and Path
- [ ] Notify PM (Fran Cross) when released
  - [ ] Fran to notify Randi Hecht
  - [x] Notify Grace K, and Michelle Dooley of the added permission to this role
  - [x] Notify Helpdesk of updated policy for Homepage Manager role
  - [ ] Fran to close the loop in the Slack thread [here](https://dsva.slack.com/archives/C52CL1PKQ/p1724255052594909?thread_ts=1723680168.114699&cid=C52CL1PKQ)
  - [x] Update related KBs

**Comment by FranECross** [comment]
There was a new role created for an education user. Dave P to look into if there are existing roles.

**Comment by mmiddaugh** [comment]
@davidmpickett I'm not sure the new editor role I mentioned during refinement is actually relevant after all, but here is the issue for reference: [Create Rates Editor Role #18451](https://github.com/department-of-veterans-affairs/va.gov-cms/issues/18451)

**Comment by davidmpickett** [comment]
**Suggested solution:**
Use the existing [Homepage Manager](https://prod.cms.va.gov/admin/people/permissions/homepage_manager) role and add the permission _Full Width Alert: Create new content_

There are only 6 [users with the Homepage Manager role](https://prod.cms.va.gov/admin/people?user=&status=1&role=homepage_manager&workbench_access_section__section=All):
- danielle.thierry@va.gov
- david.conlon@va.gov
- Michelle.Northuis@va.gov
- Randi.Hecht@va.gov
- Michelle.Northius@va.gov
- Joshua.Tuscher@va.gov

All of them aside from Josh are a Content Admin or above, so they can add any type of Content. Adding this Create permission to the role wouldn't change anyone's powers aside from Josh. AND it would only give Josh this power without granting him any other powers.

**Comment by davidmpickett** [comment]
Renaming this ticket to remove the references to Content Publisher and Sections as they are actually irrelevant. Our governance schema separates [Creation permissions](https://github.com/department-of-veterans-affairs/va.gov-team/tree/master/platform/cms/content-governance/content-creator-roles) from [Workflow/Moderation permissions](https://github.com/department-of-veterans-affairs/va.gov-team/tree/master/platform/cms/content-governance/content-moderation-roles)

**Comment by davidmpickett** [comment]
End of Sprint Update: 
- This ticket was an injection
- Waiting for final decision from @mmiddaugh (see [Slack thread](https://dsva.slack.com/archives/C52CL1PKQ/p1723680168114699)) before moving forward
- I have a PR draft ready if this direction is approved, otherwise I can come up with an alternate proposal

**Comment by davidmpickett** [comment]
Direction was approved [in slack](https://dsva.slack.com/archives/C52CL1PKQ/p1724194057531199?thread_ts=1723680168.114699&cid=C52CL1PKQ)

**Comment by jilladams** [comment]
The SPIKE part is done, but the permissions update itself isn't. That's hte only AC, so this will roll to SPrint 11 to get PR review / approval, merge and deploy https://github.com/department-of-veterans-affairs/va.gov-cms/pull/18958. 

Michelle also noted we need to make the CMS team aware of the permissions change, but not block on approval.

**Comment by jilladams** [comment]
In Sitewide cross-team sync today, Erika approved the plan to add the Create FWB permission to the Homepage Manager role. That PR is now unblocked. Erika also requested: 

1. That we share this news with Helpdesk, for their awareness when giving new users Homepage Manager role permissions in the future. (They should update a related runbook.)
2. We revise any related KBs about creating new Full Width Alerts to make sure they're still accurate

**Comment by jilladams** [comment]
Re: KBs and helpdesk awareness: 
* [How to do I create a Full-width Alert Banner?](https://prod.cms.va.gov/help/homepage-alerts-and-promo-banners/how-to-do-i-create-a-full-width-alert-banner) is already tagged as a VA.gov Homepage related article
* Created [VA.gov Homepage account administration policy](https://prod.cms.va.gov/help/cms-account-admin-policies/vagov-homepage-account-administration-policy), that spells out that we need Michelle's approval to add new people to this role, and shared with Helpdesk here:  https://dsva.slack.com/archives/CDHBKAL9W/p1724279559631059?thread_ts=1724277597.962409&cid=CDHBKAL9W

**Comment by jilladams** [comment]
Moving back to current sprint not started and unassigning Dave. 
The PR that's up is mostly in good shape, however: when logged in as Josh Tuscher, the Path field isn't editable. An engineer will need to look into why that is / what we need to adjust in terms of field/permissions to make it possible for the Homepage Manager role to not only create new FWB, but also modify the Path field.

**Comment by davidmpickett** [comment]
Because I can't help myself, further spelunking:

## [VA.gov Event subscriber](https://github.com/department-of-veterans-affairs/va.gov-cms/blob/main/docroot/modules/custom/va_gov_banner/src/EventSubscriber/EntityEventSubscriber.php#L62)

This is the place where we are disabling the Path field on Full Width Banners for non-admins.

```
  /**
   * Disable the url paths field on entity forms.
   *
   * @param array $form
   *   The entity form.
   */
  public function disablePathsField(array &$form): void {
    if (!$this->userPermsService->hasAdminRole()) {
      $form['field_target_paths']['#disabled'] = TRUE;
    }
  }
```

## [VA.gov User Perms Services](https://github.com/department-of-veterans-affairs/va.gov-cms/blob/main/docroot/modules/custom/va_gov_user/src/Service/UserPermsService.php)

- This is where we do the work to determine if the current user is an admin or not
- The function `hasAdminRole` is probably doing what it should be doing
- If we wanted a general function that could check if a user has a specific role, that would probably be a new function like hasSpecificRole which would accept an argument of the role(s) we are looking for 
- For the situation of Full Width banner and disabling the Path field we'd probably want to check if the user hasAdminRole OR hasSpecificRole = homepage_manager and disable the field if Neither of those returns true.

**Comment by dsasser** [comment]
## End of Sprint Update

We didn't get to this due to adding feature toggle work to the telephone migration.

**Comment by dsasser** [comment]
## End of Sprint Update 10-29-24

Work is in PR pending code and QA reviews. The PR was started by Dave a back in August but we needed to sort out the permissions to the 'paths' field on the Banner form, which I added [here](https://github.com/department-of-veterans-affairs/va.gov-cms/pull/18958#issuecomment-2444746868).

**Comment by jilladams** [comment]
Should have gone out in today's daily deploy. We should be able to verify as the specified user in Staging tomorrow in order to close.

**Comment by jilladams** [comment]
Nevermind! Verified today, closing: 

![Image](https://github.com/user-attachments/assets/2632f701-67a0-4c4a-a1e3-ef70ae2363cf)
